packages/gatsby/src/schema/graphql-engine/bundle-webpack.ts

```diff
@@ -38,69 +50,238 @@ function getApisToRemoveForQueryEngine(): Array<GatsbyNodeAPI> {
   return apisToRemove
 }
 
-const getInternalPackagesCacheDir = (): string =>
-  path.join(process.cwd(), `.cache/internal-packages`)
+const getInternalPackagesCacheDir = (
+  functionsTarget: IPlatformAndArch
+): string =>
+  path.posix.join(
+    slash(process.cwd()),
+    `.cache`,
+    `internal-packages`,
+    `${functionsTarget.platform}-${functionsTarget.arch}`
+  )
 
 // Create a directory and JS module where we install internally used packages
-const createInternalPackagesCacheDir = async (): Promise<void> => {
-  const cacheDir = getInternalPackagesCacheDir()
+const createInternalPackagesCacheDir = async (
+  functionsTarget: IPlatformAndArch
+): Promise<void> => {
+  const cacheDir = getInternalPackagesCacheDir(functionsTarget)
   await fs.ensureDir(cacheDir)
-  await fs.emptyDir(cacheDir)
 
   const packageJsonPath = path.join(cacheDir, `package.json`)
 
-  await fs.outputJson(packageJsonPath, {
-    name: `gatsby-internal-packages`,
-    description: `This directory contains internal packages installed by Gatsby used to comply with the current platform requirements`,
-    version: `1.0.0`,
-    private: true,
-    author: `Gatsby`,
-    license: `MIT`,
-  })
+  if (!fs.existsSync(packageJsonPath)) {
+    await fs.emptyDir(cacheDir)
+
+    await fs.outputJson(packageJsonPath, {
+      name: `gatsby-internal-packages`,
+      description: `This directory contains internal packages installed by Gatsby used to comply with the current platform requirements`,
+      version: `1.0.0`,
+      private: true,
+      author: `Gatsby`,
+      license: `MIT`,
+      functionsTarget,
+    })
+  }
 }
 
-// lmdb module with prebuilt binaries for our platform
-const lmdbPackage = `@lmdb/lmdb-${process.platform}-${process.arch}`
-
-// Detect if the prebuilt binaries for lmdb have been installed. These are installed under @lmdb and are tied to each platform/arch. We've seen instances where regular installations lack these modules because of a broken lockfile or skipping optional dependencies installs
-function installPrebuiltLmdb(): boolean {
+function getLMDBBinaryFromSiteLocation(
+  lmdbPackageName: string,
+  version: string,
+  functionsTarget: IPlatformAndArch
+): string | undefined {
   // Read lmdb's package.json, go through its optional depedencies and validate if there's a prebuilt lmdb module with a compatible binary to our platform and arch
   let packageJson: PackageJson
   try {
     const modulePath = path
-      .dirname(require.resolve(`lmdb`))
+      .dirname(slash(require.resolve(`lmdb`)))
       .replace(`/dist`, ``)
     const packageJsonPath = path.join(modulePath, `package.json`)
     packageJson = JSON.parse(fs.readFileSync(packageJsonPath, `utf-8`))
   } catch (e) {
     // If we fail to read lmdb's package.json there's bigger problems here so just skip installation
-    return false
+    return undefined
```

Comment:
`undefined` is better than a boolean value as a default to this function, but we might also be able to keep the return type strictly a string